posts/specfile-advanced/index.md

```diff
@@ -0,0 +1,252 @@
+```
+
+Doing this means that all conditions and macros will be already expanded, so all tags will be valid
+and their `value` and `expanded_value` will be the same. Modifications will have no effect.
+
+You can also access the underlying `rpm.spec` object with the `Specfile.rpm_spec` property,
+which allows you to do things like:
+
+```python
+spec = Specfile("example.spec")
+
+print(spec.rpm_spec.sourceHeader.epoch)
+```
+
+In this case the printed value will be a number, not a string like in previous cases.
+
+:::tip
+
+This means you have 3 ways how to get expanded value of a tag with the _specfile_ library.
+All of them should give you same or very similar results, but only the first one gives you access
+also to the unexpanded value of a tag and the possibility to modify it. On the other hand it will
+also take the most time, so think twice if performance is a concern.
+
+:::
+
+## Making modifications less obstructive
+
+Consider the following in a spec file:
+
+```specfile
+#global rc_ver 3
+%global maj_ver 12
+%global min_ver 0
+%global patch_ver 1
+
+Name:       %{pkg_name}
+Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
+```
+
+Rather than replacing the value of `Version` with a version string, removing all the macro
+substitutions in the process, you can use `Specfile.update_tag()` and have it try to update
+the relevant macro definitions instead:
+
+```python
+from specfile import Specfile
+
+spec = Specfile("example.spec")
+
+spec.update_tag("Version", "12.1.0")
+```
+
+This will result in:
+
+```specfile
+#global rc_ver 3
+%global maj_ver 12
+%global min_ver 1
+%global patch_ver 0
+
+Name:       %{pkg_name}
+Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
+```
+
+It will even uncomment a commented-out macro definition if needed:
+
+```python
+spec = Specfile("example.spec")
+
+spec.update_tag("Version", "12.1.1~rc1")
+```
+
+This will result in:
+
+```specfile
+%global rc_ver 1
+%global maj_ver 12
+%global min_ver 1
+%global patch_ver 1
+
+Name:       %{pkg_name}
+Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
+```
+
+However, if you want to go back to a regular version, you have to comment out the relevant
+macro definition explicitly:
+
+```python
+from specfile import Specfile
+from specfile.macro_definitions import CommentOutStyle
+
+spec = Specfile("example.spec")
+
+with spec.macro_definitions() as macro_definitions:
+    # use the original comment-out style
+    macro_definitions.rc_ver.comment_out_style = CommentOutStyle.HASH
+    macro_definitions.rc_ver.commented_out = True
+
+spec.update_tag("Version", "12.1.1")
+```
+
+This will result in:
+
+```specfile
+#global rc_ver 1
+%global maj_ver 12
+%global min_ver 1
+%global patch_ver 1
+
+Name:       %{pkg_name}
+Version:    %{maj_ver}.%{min_ver}.%{patch_ver}%{?rc_ver:~rc%{rc_ver}}
+```
+
+The algorithm isn't perfect and won't work in all situations, but even if it's not possible to update
```

Comment:
Do you think you can describe how the implementation works so people can understand how you match the macros?

Reply:
I can try, but it would be probably worth a separate blog post :slightly_smiling_face:

Reply:
I tried to describe it in https://github.com/packit/packit.dev/pull/987/commits/f594540f2d6667eaf8a653e92003e584e354fbc3. WDYT?

Reply:
Let me merge it as it is, we can always do adjustments later.